### DISCUSSION TITLE
Exactly one constraints

### DISCUSSION THREAD

**Discussion by jezberg** [SAT-STEP/SAT-STEP#54] [Ideas]
Hi 
I was mock preparing for a lecture with the program yesterday and focused on what kind of features I would still want. 

I did some ugly (local) hacks to the code in order to add the constraints "each value should appear on every row and column exactly once". This has a dramatic effect on the number of constraints learned so it would be nice to get the opportunity to solve both with and thout these constraints. That would be a strong way of demonstrating that the encoding matters (which is something people seem to not believe). I guess "each value appears in a subgrid exactly once" would also be nice. 

Note that with both "each value appears in a row exactly once" and "each value appears in a column exactly once" many of the simpler sudokus will solve without any conflicts. You can however still find examples that need conflict: 
https://abcnews.go.com/blogs/headlines/2012/06/can-you-solve-the-hardest-ever-sudoku

**Comment by ghost** [comment]
Would that be the encoding described in the original document you provided for this project
- "Each square gets at least one value" and
- "Two squares on the same row do not get the same value") etc.?

Or would this include some other constraints as well?

The current plan is to implement the originally intended encoding described above alongside the current one, and the bit encoding. Once that is done, it should be fairly easy to add different decimal encodings if needed.

**Reply by jezberg** [reply]
What I had in mind is the option to add "additional redundant" constraints with the aim of demonstrating that such constrainta matter and affect search. 

So while the bit encoding is an entirely differernt encoding, one could view variants of the current ecoding not as different, but instead augmented. 

Right now I believe the encoding is:
Each square gets exactly one value, 
No two squares on the same row/column/smallsquare get the same value. 

However, even an encoding where the first part is changed to:
Each square gets at least one value
will produce correct results. 

My proposal would be to add controll over whether the "at most one" constraints are added prior to solving. 

So there would be an option to add constraints:
Each square gets at most one value (these are currently added by default but need not be) 
Each value appears on a row/colums/small-square at most one time (these are not added).
And then solve in order to demonstrate that, while all are correct encodings in the sense that the sudoku gets solved (fast), the constraints do affect the number of extra clauses learnt during search.

**Reply by ghost** [reply]
Would there be a problem with allowing insufficient sets of constraints? I think the simplest solution with the most flexibility would be a group of checkboxes for the various constraints, with some sensible default set selected. This would mean that the results from cadical could be nonsensical depending on your selections (and that the result wouldn't necessarily be shown in full, if a square would get multiple values etc.). The other options I see are
- Having predefined sets of constraints to choose from, which limits flexibility
- Having some way for the program to distinguish a sufficient set from an insufficient one, which sounds quite difficult

**Reply by jezberg** [reply]
In principle, only thinking about someone like myself as a user, then having checkboxes for everything would be great. 

However, I do believe that (for someone else using) it would be nice to have a warning if you do not include a sufficient set of constraints. More precisely, you could warn the user if the chosen set of constraints does not include all of the following:

1) Least one value for each square
2) At most one of each value in every row / column / small box (three different sets). 

You could still allow the processing to happen anyway and show whatever, but it might be good to warn the user that the results will be nonsensical. 

Autmatically detecting specific sets of constraints that are sufficient is difficult and I don't see extra value in that from the use case of lecturing. 

Predefined sets of constraints might be good, but as you say, that limits flexibility.
